scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,82 @@
+
+## Decision
+
+We will implement a versioning stream for vulnerability bundles, bumped when we determine that ClairCore's schema has changed.
+
+Initially, two version streams will exist (table 1 below):
+
+| Stream | Git Ref      | Description                                                        |
+|--------|--------------|--------------------------------------------------------------------|
+| dev    | heads/master | This stream tracks the latest vulnerability schema                 |
+| v1     | heads/4.4.5  | This stream tracks the schema used by the current Stackrox release |
+
+On the subsequent releases (for example,. `4.4.6`), one of the two will be carried out:
+
+1.  The schema hasn't changed, so v1 now tracks the new release (`heads/4.4.6`).
+2.  The schema has changed, so v1 stays at current release (`heads/4.4.5`), and a new version, v2, is created to track the next release (`heads/4.4.6`).
+
+### Building bundle version in Exporter
+
+The GitHub workflow will have a map with "bundle version" as keys, and "git reference" as values, similar to table 1 above.  This can be stored statically in the matrix definition.
+
+For each version in the mapping, a bundle will be built based on the specified git ref and stored at `https://definitions.stackrox.io/v4/<bundle-version>/vulnerabilities.zip`.
+
+In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).  Notice that if the version is the same, it will be updated.  If it's new, it will be added, and the previous version will stay.
+
+### Selection bundle version in Scanner
+
+Scanner will determine which version bundle it wants based on hard-coded information instead during build time.  A file at `scanner/VULNERABILITY_VERSION` will determine the version on release builds and nightlies.  On non-release builds this will be set to "dev".
+
+Today scanner supports a special variable `ROX_VERSION` in the vulnerability URL configuration.  A new variable will be added called `ROX_SCANNER_VULNERABILITY_VERSION` to specify the expected bundle version to fetch.
+
+When a schema bump is detected (e.g., a ClairCore bump that brings in incompatible schema changes to the vulnerability structs), this file should be updated.
+
+The definitions handler in Central can be kept the same.  It already supports arbitrary versions in the GET request from Scanner.
+
+[The check for nightlies](https://github.com/stackrox/stackrox/blob/e332cd19a639f59d9931414a4f4e561981396dad/central/scannerdefinitions/handler/handler.go#L661) is unnecessary and will be removed.
+
+### Offline mode
+
+The offline mode will continue to serve every supported bundle version.  But instead of offering the latest Y-stream, it will offer all bundles supported by the vulnerability workflow, except for "dev".
```

Comment:
Not sure if this has been updated since our conversation earlier, if all versions are added to the latest bundle then Central will need to be updated to choose / validate the correct version after upload.  

Also after reviewing what was implemented - older versions of ACS (ie: `4.4.x` and `4.5.x`) may need to have fixes backported to accept a bundle with multiple versions, we'll have to be careful to reduce impact to existing offline workflows that use latest.

Reply:
Sorry I missed this part.  I think in offline mode we should instead offer the latest versioned bundle, which is v1 now (but in the future v2, v3, etc). If customer have trouble with accepting the latest versioned bundle, they will suggested to upgrade ACS. What do you guys think? @jvdm  @dcaravel

Reply:
> Not sure if this has been updated since our conversation earlier, if all versions are added to the latest bundle then Central will need to be updated to choose / validate the correct version after upload.

I am considering making Central serve all of them.

For previous offline bundles, I am considering converting them all to `v1` internally.

Any comments?

Reply:
I am not a fan of moving offline latest to new bundle version at every release because that would break customers until they upgrade, or they change their scripts to use `roxctl`.